src/common/AndroidUtils.ts

```diff
@@ -504,29 +529,249 @@ export class AndroidUtils {
+            } else {
+                return Promise.resolve();
+            }
+        });
+    }
+
+    /**
+     * Attempts to reboot an emulator.
+     *
+     * @param portNumber The ADB port of the emulator.
+     * @param waitForBoot Optional boolean indicating whether it should wait for the device to boot up. Defaults to true.
+     */
+    public static async rebootEmulator(
+        portNumber: number,
+        waitForBoot: boolean = true
+    ): Promise<void> {
+        return AndroidUtils.executeAdbCommand('shell reboot', portNumber).then(
+            () => {
+                if (waitForBoot) {
+                    return AndroidUtils.waitUntilDeviceIsReady(portNumber);
+                } else {
+                    return Promise.resolve();
+                }
+            }
+        );
+    }
+
+    /**
+     * Determines if an emulator was launched with -writable-system parameter by looking at its emu-launch-params.txt file.
+     *
+     * @param portNumber The ADB port of the Android virtual device.
+     * @returns A boolean indicating whether the emulator on the given port was launched with -writable-system parameter
+     */
+    public static async isEmulatorSystemWritable(
+        portNumber: number
+    ): Promise<boolean> {
+        try {
+            const emuPath = (
+                await AndroidUtils.executeAdbCommand('emu avd path', portNumber)
+            )
+                .split('\n')[0]
+                .trim();
+
+            const paramFile = path.join(emuPath, 'emu-launch-params.txt');
+
+            const launchParams = fs.readFileSync(paramFile, 'utf8');
+            return Promise.resolve(launchParams.includes('-writable-system'));
+        } catch (error) {
+            this.logger.warn(
+                `Unable to determine if emulator is system writable: ${error}`
+            );
+        }
+
+        return Promise.resolve(false);
+    }
+
+    /**
+     * Given an emulator that is running on a port, it returns the name of the AVD that is running on that port.
+     *
+     * @param portNumber The ADB port of the Android virtual device.
+     * @returns The name of the AVD that is running on that specified port.
+     */
+    public static async fetchEmulatorNameFromPort(
+        portNumber: number
+    ): Promise<string> {
+        return AndroidUtils.executeAdbCommand(
+            'emu avd name',
+            portNumber
+        ).then((result) => result.split('\n')[0].trim());
+    }
+
+    /**
+     * Remounts adb as root with writable system access for the AVD that is running on the specified port. If the AVD currently
+     * is not launched with writable system access, this function will restart it with write access first then remounts as root.
+     *
+     * @param portNumber The ADB port of the Android virtual device.
+     * @param numRetries Optional parameter that indicates the number of retries if an error occurs. Defaults to 0 (no retries).
```

Comment:
In the interest of performance I've decided to default numRetires=0 unless a caller explicitly sends a non-zero value. I'm open to changing default value to something else (like 3 tries). I thought to leave it like this first and see if we still see adb command failures in the wild before increasing this number.

Reply:
I feel like these values should be configured in code (or outside of it, if we feel that's necessary), and not otherwise passed in as parameters. It's logic internal to the method itself—it shouldn't drive the API of the method.

Reply:
@khawkins There are only 2 functions that use `numRetries` and `retryDelay`. These 2 are `executeAdbCommand` and `remountAsRootWritableSystem`.

For `remountAsRootWritableSystem` we could take the parameters out from function signature and put these inside its code. But I didn't do that b/c for `executeAdbCommand` I think it should take them as parameters. There are many functions that call into `executeAdbCommand` such as:

```
stopEmulator() // calls it to shut down an AVD
rebootEmulator() // calls it to reboot an AVD
isEmulatorSystemWritable() // calls it to get the path to the AVD folder on host machine
fetchEmulatorNameFromPort() // calls it get the name from AVD running on a port
waitUntilDeviceIsReady() // calls it to run shell on an AVD and wait for device
launchURLIntent() // calls it to open url
launchNativeApp() // calls it to install & launch an app
remountAsRootWritableSystem() // calls it to run a series of commands
```

With the exception of the last one (which is where we actually observed adb issues) none of the other ones need their commands to be retried.

So either we should have the parameters for `executeAdbCommand` OR have 2 versions of it `executeAdbCommand` and `executeAdbCommandWithRetry` and then have `executeAdbCommandWithRetry` define these inside its code instead of taking them as parameters (but then that will lead to code duplication).

Reply:
It feels like `executeAdbCommandWithRetry` *should* be its own method, calling `executeAdbCommand` internally, but managing all of the "if error message retry" logic itself. It's a subtle distinction, but it doesn't feel like `executeAdbCommand` should be carrying that burden when it's not really its responsibility, at the end of the day.

I guess where that doesn't work is the fact that `executeAdbCommand` currently encapsulates the I/O aspects of its responsibilities and only returns output.  I guess it's fine to leave it as is.